app/controllers/topics_controller.rb

```diff
@@ -84,6 +84,12 @@ def show
       end
     end
 
+    topic = Topic.where(id: (params[:id] || params[:topic_id])).first
```

Comment:
the topic is being looked up twice here, I think all the guardian checks etc should happen after we got the topic view and know the user has no access, so this only triggers on exceptional cases